nannyml/drift/univariate/methods.py

```diff
@@ -481,3 +481,80 @@ def _alert(self, data: pd.Series):
         alert = value < self.lower_threshold or value > self.upper_threshold
 
         return alert
+
+
+@MethodFactory.register(key='hellinger_distance', feature_type=FeatureType.CONTINUOUS)
+class HellingerDistance(Method):
+    """Calculates the Hellinger Distance between two distributions.
+
+    An alert will be raised for a Chunk if .
+    """
+
+    def __init__(self, **kwargs) -> None:
+        super().__init__(
+            display_name='Hellinger distance',
+            column_name='hellinger_distance',
+            **kwargs,
+        )
+
+        self.upper_threshold = 0.1
+
+        self._treat_as_type: str
+        self._bins: np.ndarray
+        self._reference_proba_in_bins: np.ndarray
+
+    def _fit(self, reference_data: pd.Series):
+        if _column_is_categorical(reference_data):
+            treat_as_type = 'cat'
+        else:
+            n_unique_values = len(np.unique(reference_data))
+            len_reference = len(reference_data)
+            if n_unique_values > 50 or n_unique_values / len_reference > 0.1:
+                treat_as_type = 'cont'
+            else:
+                treat_as_type = 'cat'
+
+        if treat_as_type == 'cont':
+            bins = np.histogram_bin_edges(reference_data, bins='doane')
+            reference_proba_in_bins = np.histogram(reference_data, bins=bins)[0] / len_reference
+            self._bins = bins
+            self._reference_proba_in_bins = reference_proba_in_bins
+        else:
+            reference_unique, reference_counts = np.unique(reference_data, return_counts=True)
+            reference_proba_per_unique = reference_counts / len(reference_data)
+            self._bins = reference_unique
+            self._reference_proba_in_bins = reference_proba_per_unique
+
+        self._treat_as_type = treat_as_type
+
+        return self
+
+    def _calculate(self, data: pd.Series):
+        reference_proba_in_bins = copy(self._reference_proba_in_bins)
+        if self._treat_as_type == 'cont':
+            len_data = len(data)
+            data_proba_in_bins = np.histogram(data, bins=self._bins)[0] / len_data
+
+        else:
+            data_unique, data_counts = np.unique(data, return_counts=True)
+            data_counts_dic = dict(zip(data_unique, data_counts))
+            data_count_on_ref_bins = [data_counts_dic[key] if key in data_counts_dic else 0 for key in self._bins]
+            data_proba_in_bins = np.array(data_count_on_ref_bins) / len(data)
+
+        leftover = 1 - np.sum(data_proba_in_bins)
```

Comment:
It would be good to add a comment here about what leftover is and why we need it.

Also, assuming I m getting this right @jakubnml are leftovers always from one side, or could the be from both (before start and after finish). Or this doesn't matter?

Reply:
For continuous data could be from both sides. For categorical data leftovers is all the unseen categories. According to the experiments we did, for both of the methods it doesn't matter if you create multiple _bins_ for chunk data and multiple _bins_ with for reference data (all with 0 relative frequency). For other methods (that we dont have in NML yet) it may be important.

Reply:
By both methods I mean both methods that use the same binning logic - JS and Hell.